### ISSUE TITLE
Ignore assets in stacks that are not being deployed

### ISSUE THREAD

**Issue by AlexCheema** [aws/aws-cdk#3418]
**Note: for support questions, please first reference our [documentation](https://docs.aws.amazon.com/cdk/api/latest), then use [Stackoverflow](https://stackoverflow.com/questions/ask?tags=aws-cdk)**. This repository's issues are intended for feature requests and bug reports.

* **I'm submitting a ...**
  - [ ] :beetle: bug report
  - [X] :rocket: feature request
  - [ ] :books: construct library gap
  - [ ] :phone: security issue or vulnerability => Please see [policy](https://github.com/aws/aws-cdk/security/policy)
  - [ ] :question: support request => Please see note at the top of this template.

* **What is the current behavior?**
*If the current behavior is a :beetle:bug:beetle:: Please provide the steps to reproduce*
Let there be an app with stacks A, B and C. Given that I deploy A and B exclusively, assets in C should not be required. Currently, if an asset in C is not found, I still cannot deploy the app, despite not wanting to deploy C at all.

* **What is the expected behavior (or behavior of feature suggested)?**
I should be able to deploy individual stacks without having assets in other stacks on disk.

* **What is the motivation / use case for changing the behavior or adding this feature?**
We have a very slow build step that creates an asset and is required for one particular stack of our app. However, if we want to deploy the app without this stack, we still have to build this asset (or put a dummy file in its place as a hacky workaround). We shouldn't have to care what happens in stacks that aren't being deployed.

* **Please tell us about your environment:**

  - CDK CLI Version: 1.1.0
  - Module Version: 1.1.0
  - OS: Windows 10
  - Language: Typescript

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. associated pull-request, stackoverflow, gitter, etc)

**Comment by SomayaB** [comment]
Hi @AlexCheema, thanks for submitting a feature request, it seems like a reasonable ask. We will look into this and someone will update this issue when there is movement.

**Comment by eladb** [comment]
Hi @AlexCheema, can you provide more details about your use case. I am not sure I understand when you ran into a situation where the source assets for parts of your app don't exist. 

Closing for now.

**Comment by SimonLdj** [comment]
Hi,

I would like to describe another use case in which such feature is vital.

### Use Case

we have 3 stacks `commonStack`, `stackA` and `stackB`. stack A and B uses resources from common stack. Stack A and B each contain different Node.js lambda function (and some sqs/sns). 
Each lambda function code written in TypeScript, bundled with Webpack and referred by the lambda using `lambda.Code.fromAsset` for example `lambda.Code.fromAsset(path.join(__dirname, '../../resources/lambdaA/.webpack'))`

For CI/CD we use github workflows detecting changes in stack definition code or in corresponding `resources/lambdaX` folder, invoking  deploy only for the effected stack. For example `.github/workflows/lambdaA-prod.yml` will run only if `resources/lambdaA` changed and will deploy `stackA` (and if needed CommonStack) only. 

As of now, cdk require all asserts files used by lambdas to exist - even for lambdas belong to a stack not being deployed.
This means when running workflow to deploy lambdaA, we still need to build/bundle lambdaB.
This mostly frustrating as it 
a) costs more when building all resources projects even only one have been changed
b) can block deploying lambdaA if lambdaB for example fails to build
all those can be worst in big mono repo with multiple teams working and deploying.

our project structure:
```
mono-git-repo
|.github/
|- workflows/
|- - lambdaA-prod.yml
|- - lambdaB-prod.yml
|cdk/
|- bin/
|- - main-cdk.ts
|- lib/
|- - commonStack.ts
|- - stackA.ts
|- - stackB.ts
|resources/
|- lambdaA/
|- - src/
|- - - index.ts
|- - .webpack/
|- - package.json
|- - tsconfig.json
|- - ...
|- lambdaB/
|- - src/
|- - - index.ts
|- - .webpack/
|- - package.json
|- - tsconfig.json
- - ...
```

### Other
Workaround we used so far: 
- building all assets in the project even when deploying specific lambda
- build and deploy all sacks even when only one of them changed

**Comment by SimonLdj** [comment]
Hi @eladb / @SoManyHs ,
could be we get some comment about this? thanks

**Comment by MariiaSokur** [comment]
Hi! I'm also interested in this question. Looking forward to hearing about possible solutions!

**Comment by fcoelho** [comment]
@eladb This is also something that interests me, but not as a bug as in the original report, more of a feature request.

My use case is how long it takes to deploy a set of stacks. I have an app that is composed of 3 stacks and to deploy them I can easily do `cdk deploy --all`. But if I'm deploying just one of them (e.g. during development) then the CDK will still try bundling all assets from all other stacks.

This is a concern for me because even if stack I'm trying to deploy is fairly small, one of the other stacks has a couple dozen assets, which means redeploying the small one takes as much time as redeploying everything. What I usually do during dev is to comment out creating this larger stack. But would be nice if the CDK could just ignore the other assets if they're not required for the deployment

**Comment by benwainwrightcinch** [comment]
Also interested in this; as @fcoelho says - its about how long it takes to deploy. I've got multiple stacks in an application, but sometimes when I'm testing I'm only deploying one of them. It's frustrating that everything has to be present, and everything has to be bundled regardless of what I'm deploying.

**Comment by vladimirmutafov** [comment]
I think this should be re-opened. We're also hitting this limitation when dealing with multiple CDK stacks. One of them is a SPA app and the other is our backend with lots of services that take quite a lot of time to build. With the current CDK versions the only way to deploy our SPA is by building all of our services as CDK tries to gather the assets for all stacks, not only the one we're deploying.

**Comment by urkle** [comment]
I just ran into this myself and it was unexpected behavior to have assests from ALL of my stacks being built when I only specified one of them to be deployed.  Having the synth phase ONLY work on the specified stack would be extremely welcome.  Having to have completely separate projects for everything just to be able to (more quickly) develop one stack is not helpful especially when I have shared framework pieces as well.
